202112 addrman unit tests

mzumsande opened the pull request:
just a fake PR

TheBlueMatt committed:
Expose block filters over REST.

This adds a new rest endpoint:
/rest/blockfilter/filtertype/requesttype/blockhash (eg
/rest/blockfilter/basic/header/000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f.hex)
which exposes either the filter "header" or the filter data itself.
Most of the code is cribbed from the equivalent RPC.

mzumsande committed:
test: Add test for updating addrman entries

This covers Connected() which updates nTime, and SetServices()
which updates nServices

amitiuttarwar committed:
test: Update addrman_serialization unit test to use AddrMan's interface

By updating the test to use FindEntry, it no longer needs to reach into
AddrMan's internals (via GetBucketAndEntry) to assert expected
behaviors.

amitiuttarwar committed:
test: delete unused GetBucketAndEntry function

mzumsande committed:
test: Add test for multiplicity in addrman new tables

TheBlueMatt committed:
Update REST docs with new accessors

glozow committed:
[doc] current rbf policy

glozow committed:
[doc] CPFP carve out and single-conflict RBF exemption

glozow committed:
[doc] clarify RBF difference from BIP125

It can be argued that BIP125#1 signaling is honored because
descendants of signaling transactions are replaceable by RBF.

Regardless, since there are multiple details in our RBF policy that are
not captured in BIP125, point to our doc instead.

amitiuttarwar committed:
addrman: Introduce a test-only function to lookup addresses

Co-Authored-By: Martin Zumsande <mzumsande@gmail.com>

mzumsande committed:
test: use AddrMan instead of AddrManTest where possible

Switches to AddrMan for tests that use no features of AddrManTest.
Also removes unusued AddrManTest variables

Co-Authored-By: Amiti Uttarwar <amiti@uttarwar.org>

mzumsande committed:
test: Inline SimConnFail function

No need for a function, since it is only used once.

mzumsande committed:
test: Remove tests for internal helper functions

The logic of these functions is already covered by existing unit tests
using publicly exposed functions of the interface.
Therefore, removing them does not decrease test coverage.

mzumsande committed:
test: Remove unused AddrManTest class

josibake committed:
refactor: addrman_selecttriedcollisions test

Check `Good()` directly when adding addresses.
Previously, test would check `size()`, which is incorrect.

Check that duplicates are also handled by checking the
output from `SelectTriedCollision()` when `Good()` returns
false.

josibake committed:
refactor: addrman_noevict test

Check the response from `Good()` wherever it is called.

Previously, the test was using `size()` (incorrect for checking tried)
and `SelectTriedCollision()` to determine if a collision happened.

josibake committed:
refactor: addrman_evictionworks test

Test for collisions and duplicates directly with `Good()`.

If an entry to tried is a duplicate, `Good()` will return false
but `SelectTriedCollision()` will be empty (assuming there were no prior
collisions). If there is a collision, `Good()` will retun false
and `SelectTriedCollision()` will return a value.

josibake committed:
refactor: addrman_select test

Check that `Good()` is successful whenever it is called.

mzumsande committed:
test: Cover eviction by timeout in addrman_evictionworks

jnewbery commented on src/addrman.h:
```diff
@@ -22,6 +22,29 @@ class AddrManImpl;
 
+/** Test-only struct, capturing info about an address in AddrMan */
+struct AddressPosition {
+    // Whether the address is on the new or tried table
```
s/address is on/address is in/

jnewbery commented on src/addrman.h:
```diff
@@ -22,6 +22,29 @@ class AddrManImpl;
+    // Whether the address is on the new or tried table
+    bool tried{false};
+
+    // Addresses on the tried table should always have a multiplicity of 1.
```
s/on/in/ (same below)

jnewbery commented on src/addrman.h:
```diff
@@ -22,6 +22,29 @@ class AddrManImpl;
+    // ADDRMAN_NEW_BUCKETS_PER_ADDRESS
+    int multiplicity{0};
+
+    // If the address is on the new table, bucket & position should be
```
If the address is in the new table, the bucket and position are

jnewbery commented on src/addrman.h:
```diff
@@ -136,6 +159,15 @@ class AddrMan
+    /*
+     * Test-only function
+     *
+     * Finds the address record in AddrMan and returns information about its
```
Use the indicative mood:

Find the address record in AddrMan and return information about its

jnewbery commented on src/addrman.h:
```diff
@@ -136,6 +159,15 @@ class AddrMan
+     * position(s). If the address is not found, returns an empty
+     * AddressPosition struct with multiplicity set to 0.
+     */
+    AddressPosition FindAddressEntry(const CAddress& addr);
```
I think it'd be better to return a std::optional<AddressPosition> and std::nullopt if the address isn't found, rather than a null AddressPosition.

jnewbery commented on src/test/addrman_tests.cpp:
```diff
@@ -151,232 +98,192 @@ BOOST_AUTO_TEST_CASE(addrman_simple)
+    BOOST_CHECK(addrman->Add({CAddress(addr2, NODE_NONE)}, source));
+    BOOST_CHECK_EQUAL(addrman->size(), num_addrs - collisions);
+}
+
+
```
```suggestion

 
```

jnewbery commented on src/test/addrman_tests.cpp:
```diff
@@ -107,7 +54,7 @@ BOOST_FIXTURE_TEST_SUITE(addrman_tests, BasicTestingSetup)
 BOOST_AUTO_TEST_CASE(addrman_simple)
 {
-    auto addrman = std::make_unique<AddrManTest>();
+    auto addrman = std::make_unique<AddrMan>(std::vector<bool>(), /*deterministic=*/true, /*consistency_check_ratio=*/100);
```
`std::make_unique<AddrMan>(std::vector<bool>(), /*deterministic=*/true, /*consistency_check_ratio=*/100)` occurs many times in this test. Would it make sense to define a function that returns this type of addrman?

achow101 committed:
contrib: add achow101 to trusted keys

For maintaining the wallet and psbt things

shaavan committed:
Change time variable type to std::chrono::seconds in src/net_processing.cpp

- This commit is a followup to commit: 60b579
- This changes the remaining time variable in net_processing.cpp from
int64_t to std::chrono

mzumsande committed:
test: fix intermittent timeouts in p2p_timeouts.py

by checking that all nodes are added before the mocktime is bumped.
Fixes #23800

fanquake committed:
Merge bitcoin/bitcoin#23798: contrib: add achow101 to trusted keys

33b511251acbd0ed802a8843361c7dc8d1512008 contrib: add achow101 to trusted keys (Andrew Chow)

Pull request description:

  For maintaining the wallet and psbt things

ACKs for top commit:
  glozow:
    concept ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  michaelfolkson:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  kiminuo:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  jb55:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  gruve-p:
    ACK https://github.com/bitcoin/bitcoin/pull/23798/commits/33b511251acbd0ed802a8843361c7dc8d1512008
  prayank23:
    ACK https://github.com/bitcoin/bitcoin/pull/23798/commits/33b511251acbd0ed802a8843361c7dc8d1512008
  meshcollider:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  darosior:
    Concept ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  shaavan:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  benthecarman:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  hebasto:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008
  Zero-1729:
    ACK 33b511251acbd0ed802a8843361c7dc8d1512008

Tree-SHA512: 2e242b5f763a3d9fab00413760973a41a41c78f327f4a8f7bed788d6dd9f31043a644ff5f612ff033f8a3c7b4e9936737475fec67536935257053a9f6a3fb9bb

Commit:
Merge bitcoin/bitcoin#17631: Expose block filters over REST

2b64fa3251ac5ff4b4d174f1f0be7226490dce87 Update REST docs with new accessors (Matt Corallo)
ef7c8228fd5cf45526518ae2bd5ebdd483e65525 Expose block filters over REST. (Matt Corallo)

Pull request description:

  This adds a new rest endpoint:
  /rest/blockfilter/filtertype/requesttype/blockhash (eg
  /rest/blockfilter/basic/header/000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f.hex)
  which exposes either the filter "header" or the filter data itself.
  Most of the code is cribbed from the equivalent RPC.

  You can test it at http://bitcoin-rest.bitcoin.ninja/rest//blockfilter/basic/header/000000005b7a58a939b2636f61fa4ddd62258c5fed57667a35d23f2334c4f86d.hex

ACKs for top commit:
  dergoegge:
    ACK 2b64fa3251ac5ff4b4d174f1f0be7226490dce87 - Adding blockfilters to the REST interface is analogous to serving other public data such as transactions or blocks.

Tree-SHA512: d487bc694266375c94d6fcf2e9d788a8a42a3b94e8d3290e46335a64cbcde55084ce5ea6119b79a4065888d94d7c3ae25a59a901fa46e3711f0eb296add12696

Commit:
Merge bitcoin/bitcoin#23812: test: fix intermittent failures in p2p_timeouts.py

0a1b6fa5a18f3efb2ac3e28a23a4fd5e1cf9eaf0 test: fix intermittent timeouts in p2p_timeouts.py (Martin Zumsande)

Pull request description:

  Fixes  #23800 by making sure that all peers are connected (i.e. `m_connected` is set) before the mocktime is bumped.
  We can't wait for verack here, but we can wait for a debug log entry ("Added connection peer=2") instead.

  In the failed CI runs (e.g. https://cirrus-ci.com/task/5600553806856192?logs=ci#L7469)  different peers were added at different mocktimes.

ACKs for top commit:
  naumenkogs:
    ACK 0a1b6fa5a18f3efb2ac3e28a23a4fd5e1cf9eaf0
  theStack:
    Concept and approach ACK 0a1b6fa5a18f3efb2ac3e28a23a4fd5e1cf9eaf0

Tree-SHA512: 1a3c8a9a79339d4adc6ecb1731eb0d0eadb2e5024ad3c6779b4696691f85d6c3304ef8689746d0332150a4cf04489ca4b2ff3eeb0bb76feec28c1e4bb9dbca19

Commit:
Merge bitcoin/bitcoin#23780: refactor, test: update `addrman_tests.cpp` to use output from `AddrMan::Good()`

bf4f8171352e9b384b42c91da61dfc9c3ca89ed8 refactor: addrman_select test (josibake)
5a64dc018c04ce16202a8e58ce92d2657c0b1806 refactor: addrman_evictionworks test (josibake)
e281fccd8a80d7cd48c3b17d58fd4a8915e1e965 refactor: addrman_noevict test (josibake)
8bdd9240d4310aafa1332159355f106a8fcfc5c9 refactor: addrman_selecttriedcollisions test (josibake)

Pull request description:

  As a follow-up to #23713 , this PR refactors the remaining tests in `src/tests/addrman_tests.cpp` to use the output from `AddrMan::Good()` where appropriate.

ACKs for top commit:
  naumenkogs:
    ACK bf4f8171352e9b384b42c91da61dfc9c3ca89ed8
  mzumsande:
    Code Review ACK bf4f8171352e9b384b42c91da61dfc9c3ca89ed8

Tree-SHA512: 93cc127aecff42c1c174daa04911af7e3460a5c40ddf96952fe4a6ab86fa1ff22d66724326abb709008d7f9f79c26c55c6d62753c40059c9ac60f869507ec913

Commit:
Merge bitcoin/bitcoin#23711: docs: RBF policy and mempool limit exemptions

82858bab64274506cfcd365a6588a1a9141fb22c [doc] CPFP carve out and single-conflict RBF exemption (glozow)
1fd49eb498c75a1d14193bb736d195a3dc75ae12 [doc] clarify RBF difference from BIP125 (glozow)
919ae8b8cdeccfc04026293153b876e27469a027 [doc] current rbf policy (glozow)

Pull request description:

  Since RBF was first implemented and BIP125 was written, our code has changed, people have highlighted implementation differences, and some people have proposed further changes to it. Many people seem to support the idea of documenting our _current_ RBF policy as it stands today.

  As the ancestor/descendant limit carve-out exemptions are very related to RBF, it seemed appropriate to group them with this PR.

  Related to #22806 - it seems that these policies are the most confusing for people, or at least the most documentation-requested.

ACKs for top commit:
  dunxen:
    ACK 82858ba
  t-bast:
    ACK https://github.com/bitcoin/bitcoin/pull/23711/commits/82858bab64274506cfcd365a6588a1a9141fb22c, thanks @glozow!
  darosior:
    re-ACK 82858bab64274506cfcd365a6588a1a9141fb22c
  ariard:
    ACK 82858ba

Tree-SHA512: 5d296537cce3488c18179c0aa76c739ca02fdc424e5aa17129b4cdd0d057358f86bcc1e92a9857bd2c60495f834fe9d9406d1a9f8ac5cfc8f0f4f4c27ec4f8e1

Commit:
Merge bitcoin/bitcoin#23801: Refactor: Change time variable type from int64_t to std::chrono::seconds in net_processing.cpp

92082ea0bb62ab6ec723b57ca4f2b2c5f6b28469 Change time variable type to std::chrono::seconds in src/net_processing.cpp (Shashwat)

Pull request description:

  - This is a follow-up to PR #23758
  - This changes the remaining time variable in `net_processing.cpp` from **int64_t** to **std::chrono::seconds**

ACKs for top commit:
  naumenkogs:
    ACK 92082ea0bb62ab6ec723b57ca4f2b2c5f6b28469
  hebasto:
    re-ACK 92082ea0bb62ab6ec723b57ca4f2b2c5f6b28469

Tree-SHA512: 559e351d9046d4ba2b842ae38da13b4befc7feee71f0762f97907812471e2840b0d43c90c92222d15619fe40cc21f11d40900500ca07b470f7ac8b0046cc1d68
